ironfish-zkp/src/circuits/create_asset.rs

```diff
@@ -0,0 +1,236 @@
+
+            // Compute nk = [nsk] ProvingPublicKey
+            nk = ecc::fixed_base_multiplication(
+                cs.namespace(|| "computation of nk"),
+                &PROOF_GENERATION_KEY_GENERATOR,
+                &nsk,
+            )?;
+        }
+
+        // This is the "viewing key" preimage for CRH^ivk
+        let mut ivk_preimage = vec![];
+
+        // Place ak in the preimage for CRH^ivk
+        ivk_preimage.extend(ak.repr(cs.namespace(|| "representation of ak"))?);
+
+        // Extend ivk and nf preimages with the representation of
+        // nk.
+        {
+            let repr_nk = nk.repr(cs.namespace(|| "representation of nk"))?;
+
+            ivk_preimage.extend(repr_nk.iter().cloned());
+        }
+
+        assert_eq!(ivk_preimage.len(), 512);
+
+        // Compute the incoming viewing key ivk
+        let mut ivk = blake2s::blake2s(
+            cs.namespace(|| "computation of ivk"),
+            &ivk_preimage,
+            CRH_IVK_PERSONALIZATION,
+        )?;
+
+        // drop_5 to ensure it's in the field
+        ivk.truncate(jubjub::Fr::CAPACITY as usize);
+
+        // Witness g_d, checking that it's on the curve.
+        let g_d = {
+            ecc::EdwardsPoint::witness(
+                cs.namespace(|| "witness g_d"),
+                self.owner
+                    .as_ref()
+                    .and_then(|a| a.g_d().map(jubjub::ExtendedPoint::from)),
+            )?
+        };
+
+        // Check that g_d is not small order.
+        g_d.assert_not_small_order(cs.namespace(|| "g_d not small order"))?;
+
+        // Compute pk_d = g_d^ivk
+        let pk_d = g_d.mul(cs.namespace(|| "compute pk_d"), &ivk)?;
+
+        // Hash the Asset Info pre-image
+        let identifier_preimage = hash_asset_to_preimage(
+            &mut cs.namespace(|| "asset info preimage"),
+            self.name,
+            self.chain,
+            self.network,
+            g_d,
+            pk_d,
+            self.nonce,
+        )?;
+
+        // Computed identifier bits from the given asset info
+        let asset_identifier = blake2s::blake2s(
+            cs.namespace(|| "blake2s(asset info)"),
+            &identifier_preimage,
+            ASSET_IDENTIFIER_PERSONALIZATION,
+        )?;
+
+        // Ensure the pre-image of the generator is 32 bytes
+        assert_eq!(asset_identifier.len(), 256);
+
+        multipack::pack_into_inputs(cs.namespace(|| "pack identifier"), &asset_identifier)?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::slice;
+
+    use bellman::{gadgets::multipack, groth16};
+    use bls12_381::Bls12;
+    use ff::Field;
+    use group::{Group, GroupEncoding};
+    use rand::{rngs::StdRng, SeedableRng};
+    use zcash_primitives::sapling::{Diversifier, ProofGenerationKey};
+
+    use crate::constants::{ASSET_IDENTIFIER_LENGTH, ASSET_IDENTIFIER_PERSONALIZATION};
+
+    use super::CreateAsset;
+
+    #[test]
+    fn test_create_asset_circuit() {
```

Comment:
Is it worth creating a few un-happy path tests to ensure it fails for erroneous conditions? I'm not currently sure what those would be, but potentially worth thinking on.

Reply:
Definitely. This is the kind of logic we want to be sure is pretty bulletproof. We're still getting the foundation laid so to speak, so we'll be iterating on this quite a bit, but it's on the list to add more tests